### PR TITLE
Annotate strand information for bam consensus reads

### PR DESCRIPTION
**Pull request by FelixMoelder**
This PR implements annotation of strand information for consensus reads in fastq description.
This is required for correct interpretation by varlociraptor as merged reads get mapped to the same strand and there for are identified as artifact.

### CODE REVIEW

File: src/bam/collapse_reads_to_fragments/calc_consensus.rs [hunk 1]
```diff
@@ -79,6 +87,41 @@ impl<'a> CalcOverlappingConsensus<'a> {
...
+                if rec.base(rev_base_pos) == ref_base {
+                    match rec.is_reverse() {
+                        true => strands.insert(b'-'),
+                        false => strands.insert(b'+'),
+                    };
+                }
+            });
+        }
+        match strands.len() == 1 {
+            true => consensus_strand.push(strands.take(&(b'-')).unwrap_or(b'+')),
```
Comment (FelixMoelder) [inline src/bam/collapse_reads_to_fragments/calc_consensus.rs:121]: I don't like the way how the single element from within the HashSet is accessed, but I did not find another way like poping or indexing. As it is guaranteed to contain just a single element is there any way to returning it without this `take` and `unwrap_or` thingy?

File: src/bam/collapse_reads_to_fragments/calc_consensus.rs [hunk 2]
```diff
@@ -79,6 +87,41 @@ impl<'a> CalcOverlappingConsensus<'a> {
     fn overlap(&self) -> usize {
         self.overlap
     }
+    fn build_consensus_strand(
+        &self,
+        consensus_strand: &mut Vec<u8>,
+        ref_base: u8,
+        base_pos: usize,
+    ) {
+        let mut strands = HashSet::new();
```
Comment (johanneskoester) [inline src/bam/collapse_reads_to_fragments/calc_consensus.rs:96]: I'd rather have an enum for this:

```
use std::ops::BitOrAssign;

#[derive(Eq, PartialEq)]
enum StrandObservation {
    None,
    Forward,
    Reverse,
    Both,
}

impl BitOrAssign for StrandObservation {
    fn bitor_assign(&mut self, rhs: Self) {
        if let StrandObservation::None = self {
            *self = rhs;
        } else if *self != rhs {
            *self = StrandObservation::Both;
        }
    }
}
```

Comment (johanneskoester) [reply src/bam/collapse_reads_to_fragments/calc_consensus.rs:96]: Then, you can just initialize with `let mut strand = StrandObservation::None`, and then update with `strand |= obs`.

File: src/bam/collapse_reads_to_fragments/calc_consensus.rs [hunk 3]
```diff
@@ -79,6 +87,41 @@ impl<'a> CalcOverlappingConsensus<'a> {
...
+        let fwd_end_pos = self.recs1()[0].len();
```
Comment (johanneskoester) [inline src/bam/collapse_reads_to_fragments/calc_consensus.rs:97]: Calling this fwd is a bis misleading. I'd rather say first and recond read or so.

File: src/bam/collapse_reads_to_fragments/calc_consensus.rs (hunk 1 above)
Comment (johanneskoester) [reply src/bam/collapse_reads_to_fragments/calc_consensus.rs:121]: See the enum suggestion above. This avoids sets entirely.

File: src/bam/collapse_reads_to_fragments/calc_consensus.rs [hunk 4]
```diff
@@ -162,14 +205,39 @@ impl<'a> CalcNonOverlappingConsensus<'a> {
...
     }
     pub fn recs(&self) -> &[bam::Record] {
         self.recs
     }
+    fn build_consensus_strand(
+        &self,
+        consensus_strand: &mut Vec<u8>,
+        ref_base: u8,
+        current_pos: usize,
+    ) {
```
Comment (johanneskoester) [inline src/bam/collapse_reads_to_fragments/calc_consensus.rs:226]: See my comments above.

**Review by johanneskoester — CHANGES_REQUESTED** [review]

**Review by johanneskoester — APPROVED** [review]

**Pull request merged**